// langchain/dist/callbacks/handlers/tracer_langchain.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangChainTracer = void 0;
const async_caller_js_1 = require("../../util/async_caller.cjs");
const env_js_1 = require("../../util/env.cjs");
const tracer_js_1 = require("./tracer.cjs");
class LangChainTracer extends tracer_js_1.BaseTracer {
    constructor({ exampleId, sessionName, callerParams, timeout, } = {}) {
        super();
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "langchain_tracer"
        });
        Object.defineProperty(this, "endpoint", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (0, env_js_1.getEnvironmentVariable)("LANGCHAIN_ENDPOINT") || "http://localhost:1984"
        });
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {
                "Content-Type": "application/json",
            }
        });
        Object.defineProperty(this, "sessionName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "exampleId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "caller", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "timeout", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 5000
        });
        const apiKey = (0, env_js_1.getEnvironmentVariable)("LANGCHAIN_API_KEY");
        if (apiKey) {
            this.headers["x-api-key"] = apiKey;
        }
        this.sessionName =
            sessionName ?? (0, env_js_1.getEnvironmentVariable)("LANGCHAIN_SESSION");
        this.exampleId = exampleId;
        this.timeout = timeout ?? this.timeout;
        this.caller = new async_caller_js_1.AsyncCaller(callerParams ?? { maxRetries: 2 });
    }
    async _convertToCreate(run, example_id = undefined) {
        const runExtra = run.extra ?? {};
        runExtra.runtime = await (0, env_js_1.getRuntimeEnvironment)();
        const persistedRun = {
            id: run.id,
            name: run.name,
            start_time: run.start_time,
            end_time: run.end_time,
            run_type: run.run_type,
            reference_example_id: example_id,
            extra: runExtra,
            execution_order: run.execution_order,
            serialized: run.serialized,
            error: run.error,
            inputs: run.inputs,
            outputs: run.outputs ?? {},
            session_name: this.sessionName,
            child_runs: await Promise.all(run.child_runs.map((child_run) => this._convertToCreate(child_run))),
        };
        return persistedRun;
    }
    async persistRun(run) {
        const persistedRun = await this._convertToCreate(run, this.exampleId);
        const endpoint = `${this.endpoint}/runs`;
        const response = await this.caller.call(fetch, endpoint, {
            method: "POST",
            headers: this.headers,
            body: JSON.stringify(persistedRun),
            signal: AbortSignal.timeout(this.timeout),
        });
        // consume the response body to release the connection
        // https://undici.nodejs.org/#/?id=garbage-collection
        const body = await response.text();
        if (!response.ok) {
            throw new Error(`Failed to persist run: ${response.status} ${response.statusText} ${body}`);
        }
    }
}
exports.LangChainTracer = LangChainTracer;
